src/storage/UpdateVertexProcessor.cpp

```diff
@@ -0,0 +1,344 @@
+                                            nebula::cpp2::SupportedType::BOOL);
+                   break;
+               }
+               case VAR_STR: {
+                   writer << boost::get<std::string>(v);
+                   column = this->columnDef(std::string("anonymous"),
+                                            nebula::cpp2::SupportedType::STRING);
+                   break;
+               }
+               default: {
+                   LOG(FATAL) << "Unknown VariantType: " << v.which();
+                   return;
+               }
+           }
+           respScheam.columns.emplace_back(std::move(column));
+        }
+        resp_.set_schema(std::move(respScheam));
+        resp_.set_data(writer.encode());
+    }
+}
+
+
+kvstore::ResultCode UpdateVertexProcessor::collectVertexProps(
+                            const PartitionID partId,
+                            const VertexID vId,
+                            const TagID tagId,
+                            const std::vector<PropContext>& props) {
+    auto prefix = NebulaKeyUtils::prefix(partId, vId, tagId);
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = this->kvstore_->prefix(this->spaceId_, partId, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        VLOG(3) << "Error! ret = " << static_cast<int32_t>(ret)
+                << ", spaceId " << this->spaceId_;
+        return ret;
+    }
+
+    if (iter && iter->valid()) {
+        auto reader = RowReader::getTagPropReader(this->schemaMan_,
+                                                  iter->val(),
+                                                  this->spaceId_,
+                                                  tagId);
+        const auto constSchema = reader->getSchema();
+        for (auto& prop : props) {
+            auto res = RowReader::getPropByName(reader.get(), prop.prop_.name);
+            if (!ok(res)) {
+                VLOG(1) << "Skip the bad value for tag: " << tagId
+                        << ", prop " << prop.prop_.name;
+                return kvstore::ResultCode::ERR_UNKNOWN;
+            }
+            auto&& v = value(std::move(res));
+            tagFilters_.emplace(std::make_pair(tagId, prop.prop_.name), v);
+        }
+        if (updateTagIds_.find(tagId) != updateTagIds_.end()) {
+            auto schema = std::const_pointer_cast<meta::SchemaProviderIf>(constSchema);
+            auto updater =
+                std::unique_ptr<RowUpdater>(new RowUpdater(std::move(reader), schema));
+            tagUpdaters_[tagId] = std::make_unique<KeyUpdaterPair>();
+            auto& tagUpdater = tagUpdaters_[tagId];
+            tagUpdater->key = iter->key().toString();
+            tagUpdater->updater = std::move(updater);
+        }
+    } else if (insertable_ && updateTagIds_.find(tagId) != updateTagIds_.end()) {
+        resp_.set_upsert(true);
+        const auto constSchema = this->schemaMan_->getTagSchema(this->spaceId_, tagId);
```

Comment:
becareful about constSchema, maybe it is a nullptr

Reply:
done!